ttnn/cpp/ttnn/operations/conv/conv2d/prepare_conv2d_weights.cpp

```diff
@@ -0,0 +1,367 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "conv2d.hpp"
+#include <sys/types.h>
+#include <cstdint>
+
+using namespace tt;
+namespace ttnn {
+namespace operations::conv {
+using sliding_window::SlidingWindowConfig;
+using sliding_window::ParallelConfig;
+
+namespace conv2d {
+
+void validate_weight_tensor(const ttnn::Tensor& weight_tensor) {
+    TT_ASSERT(!ttnn::has_storage_type_of(weight_tensor, ttnn::DEVICE_STORAGE_TYPE));
+    TT_ASSERT(weight_tensor.get_layout() == Layout::ROW_MAJOR);
+    TT_ASSERT(weight_tensor.get_shape().rank() == 4);
+}
+
+void validate_bias_tensor(const ttnn::Tensor& bias_tensor) {
+    TT_ASSERT(!ttnn::has_storage_type_of(bias_tensor, ttnn::DEVICE_STORAGE_TYPE));
+    TT_ASSERT(bias_tensor.get_shape().rank() == 4);
+    TT_ASSERT(bias_tensor.get_layout() == Layout::ROW_MAJOR);
+}
+
+void validate_weights_format(std::string weights_format) {
+    TT_FATAL(weights_format.size() == 4, "weights_format must have exactly 4 characters");
+    TT_ASSERT(weights_format.find("O") != string::npos, "weights_format must contain \"O\"");
+    TT_ASSERT(weights_format.find("I") != string::npos, "weights_format must contain \"I\"");
+    TT_ASSERT(weights_format.find("H") != string::npos, "weights_format must contain \"H\"");
+    TT_ASSERT(weights_format.find("W") != string::npos, "weights_format must contain \"W\"");
+    TT_ASSERT(weights_format == "OIHW", "Conv2d weights format must be \"OIHW\"");
+}
+
+template <typename T>
+OptimizedConvBlockConfig get_opt_block_config(
+    bool mm_conv,
+    uint32_t in_channels,
+    uint32_t out_channels,
+    uint32_t output_height,
+    uint32_t output_width,
+    uint32_t batch_size,
+    uint32_t input_width,
+    std::array<uint32_t, 2> kernel_size,
+    std::array<uint32_t, 2> stride,
+    T *device,
+    Conv2dConfig& conv_config) {
+
+    if (!conv_config.shard_layout.has_value()) {
```

Comment:
we still have the same problem here.
in conv2d() we run this code only if input tensor is not sharded and user hasn't specified the shard spec in Conv2dConfig.

With this change following problem could happen.

User is using conv2d with sharded tensor as input, and hasn't specified the shard spec in Conv2dConfig (which should be fine).
Before that user used prepare_conv_weights() with same Conv2dConfig (so not shard spec declared). prepare_conv_weights() will deduce the shard spec 
using auto shard codepath, and that shard spec could be different to shard spec in input tensor. prepare_conv_weights() will prep that weights for block sharding for example.
When conv2d() gets to run it will read the shard spec from the input tensor and lets say it's height sharded.
Than we will try to run height sharded codepath for conv2d with weights prepared for blocks sharding and than....

So this is a problem we need to keep these two code paths in sync, and in order to do that we need info about input_tensor sharding here.

Reply:
I've now required that the input tensor memory config is passed to the prepare functions.